Knative ConfigMap Propagation

grac3gao opened the pull request:
Fixes #1758

## Proposed Changes

- This PR implements the design from @grantr in [Knative ConfigMap Propagation](https://docs.google.com/document/d/19kubGE8RQWlPFxz38G7hAtcqfihHosXCr6zXnWWOfOI/edit#heading=h.nk7a5p2r1pvz)
- It includes two parts: 1) Creating a namespaced CRD `ConfigMapPropagation` and a corresponding controller to copy ConfigMaps from one namespace to another, 2) Updating Eventing Broker to adapt this CRD

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note

```

mattmoor commented on pkg/apis/configs/register.go:
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_types.go:
```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_validation.go:
```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/register.go:
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/register_test.go:
```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/reconciler/testing/configmap.go:
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

mattmoor commented on pkg/reconciler/testing/configmappropagation.go:
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2019 The Knative Authors
```

lionelvillard commented:
Is this something Knative Serving also needs?

grac3gao commented:
ready for review @grantr

yolocs commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,318 @@
+			return fmt.Errorf(`unable to update ConfigMap in current namespace, ConfigMap doesn't have "knative.dev/eventing/config-propagation:copy" label`)
+		}
+		if current, err = r.KubeClientSet.CoreV1().ConfigMaps(expected.Namespace).Update(expected); err != nil {
+			return fmt.Errorf("error updating ConfigMap in current namespace: %v", err)
```
You can now use `%w` verb for `fmt.Errorf`. And everywhere else.

yolocs commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,318 @@
+}
+
+func (r *Reconciler) deleteOrKeepConfigMap(ctx context.Context, cmp *v1alpha1.ConfigMapPropagation, copyConfigMap *corev1.ConfigMap, originalConfigMapName string) error {
+	originalConfigMap, err := r.configMapLister.ConfigMaps(cmp.Spec.OriginalNamespace).Get(originalConfigMapName)
```
IIUC, this is kind of duplicated logic. In `reconcileConfigMap` you already list all the configmaps from the original namespace with propagation labels. So any configmap owned by the controller in the current namespace but not in the `originalConfigMapList` should be deleted. Is that right?

grac3gao replied:
This method is for the case if the `originalConfigMap` exists before but no longer here now, then the corresponding `copyConfigMap` should be deleted. According to my understanding, from `originalConfigMap` side, it can only propagate existing ConfigMap, but it can't know which `originalConfigMap` is deleted and propagate this already deleted `originalConfigMap`. So I check it from the `copyConfigMap` side. Do you think there is a better way to make that deletion?

yolocs replied:
On L207 you select ConfigMaps from the original namespace to be `originalConfigMapList `. I think the logic for the current namespace could be simple as "if a managed ConfigMap is in the list, then update it; otherwise, delete it". Did I misunderstand something?

yolocs commented on pkg/reconciler/configmappropagation/resources/configmap_test.go:
```diff
@@ -0,0 +1,77 @@
+	configsv1alpha1 "knative.dev/eventing/pkg/apis/configs/v1alpha1"
+)
+
+func TestMakeMakeConfigMap(t *testing.T) {
```
Double 'Make'?

yolocs commented on test/common/creation.go:
```diff
@@ -395,20 +407,20 @@ func (client *Client) CreateRBACResourcesForBrokers() {
+	//	crConfigReaderName,
+	//	fmt.Sprintf("%s-%s-%s", saFilterName, helpers.MakeK8sNamePrefix(client.Namespace), crConfigReaderName),
+	//	resources.SystemNamespace,
+	//)
```
Remove?

grac3gao commented on pkg/apis/configs/v1alpha1/configmappropagation_defaults.go:
```diff
@@ -0,0 +1,29 @@
+	// If we haven't configured the original namespace,
+	// then set the default original namespace to the knative-eventing.
+	if cmp != nil && cmp.Spec.OriginalNamespace == "" {
+		cmp.Spec.OriginalNamespace = "knative-eventing"
```
I will put a TODO for that to notify.

grac3gao commented:
Done. Ready for review again.

antoineco commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,314 @@
+
+const (
+	// Name of the corev1.Events emitted from the reconciliation process
+	configMapPropagationReconcileError                  = "configMapPropagationReconcileError"
```
The value should start with a capital "C".

antoineco commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,314 @@
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Info("ConfigMapPropagation key in work queue no longer exists")
+		return nil
+	} else if err != nil {
```
(nit) a switch may be more idiomatic here
```golang
switch {
case apierrs.IsNotFound(err):
    // ...
case err != nil:
    // ...
}
```

grac3gao replied:
For this part, I followed the same way as other reconcilers did (like pkg/reconciler/broker/broker.go, pkg/reconciler/trigger/trigger.go).

antoineco replied:
Is that really a reason not to write idiomatic code where possible ?

grac3gao replied:
I am just wondering if that is more idiomatic, why no other places do that. Do you have some readability review guidelines for this?

antoineco replied:
Not at all, it may be completely arbitrary. I'm just used to seeing apierrors handled that way in the Kubernetes codebase, and I'm probably alergic to `else if`s :) Feel free to discard.

grantr replied:
@antoineco There are multiple idioms. This one came from the original kubernetes sample controller years ago (which in the meantime has [changed a bit](https://github.com/kubernetes/sample-controller/blob/bd4ff3d8adf21a05e468f522a14fabc7c4e70f65/controller.go#L252)).

If you're interested in using switches here, that's fair, but we should use switches everywhere this code is duplicated. Create an issue in eventing and serving explaining why you want to change it and create PRs.

Another option is to talk to @n3wscott who IIUC is working on generating this code. You could get your preferred idiom into the generator template :smile:

antoineco commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle.go:
```diff
@@ -0,0 +1,54 @@
+	configMapPropagationCondSet.Manage(cmps).InitializeConditions()
+}
+
+func (cmps *ConfigMapPropagationStatus) MarkConfigMapPropagationPropagated() {
```
(nit) I'm wondering whether it's really necessary to repeat "ConfigMapPropagation" inside those method names, since the method applies to a "ConfigMapPropagationStatus" already.

antoineco commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle_test.go:
```diff
@@ -0,0 +1,205 @@
+)
+
+var (
+	trueValue  = true
```
`knative.dev/pkg` has a helper for this: [ptr.Bool](https://github.com/knative/pkg/blob/master/ptr/ptr.go#L35)

antoineco commented on pkg/apis/configs/v1alpha1/configmappropagation_defaults.go:
```diff
@@ -0,0 +1,31 @@
+	// TODO if this API starts to become commonly adopted by others,
+	//  it will stop make sense defaulting to knative-eventing.
+	if cmp != nil && cmp.Spec.OriginalNamespace == "" {
+		cmp.Spec.OriginalNamespace = "knative-eventing"
```
I know it was already said, but Serving will also need this controller. Do we really want to set that default (Eventing)?

grac3gao replied:
Removed

antoineco commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle.go:
```diff
@@ -0,0 +1,54 @@
+	// ConfigMapPropagationConditionReady has status True when all subconditions below have been set to True.
+	ConfigMapPropagationConditionReady = apis.ConditionReady
+	// ConfigMapPropagationConditionPropagated has status True when the ConfigMaps in original namespace are all propagated to target namespace.
+	ConfigMapPropagationConditionPropagated apis.ConditionType = "ConfigMapPropagated"
```
Does the condition need to include "ConfigMap", since it applies to a "ConfigMapPropagation"?

grac3gao replied:
Could you explain more about this? (what kind of "ConfigMap" condition do you think that can be included)

antoineco replied:
I'm just talking about the string, I was just suggesting that "Propagated" would be less redundant "ConfigMapPropagated".

antoineco commented on pkg/reconciler/configmappropagation/resources/configmap.go:
```diff
@@ -0,0 +1,59 @@
+	}
+}
+
+func MakeCopyConfigMapName(configMapPropagationName string, configMapName string) string {
```
```suggestion
func MakeCopyConfigMapName(configMapPropagationName, configMapName string) string {
```

antoineco commented on pkg/reconciler/configmappropagation/resources/configmap.go:
```diff
@@ -0,0 +1,59 @@
+	return configMapPropagationName + "-" + configMapName
+}
+
+func MakeCopyConfigMapLabel(configMapPropagationNamespace string, configMapName string) string {
```
```suggestion
func MakeCopyConfigMapLabel(configMapPropagationNamespace, configMapName string) string {
```

antoineco commented on pkg/reconciler/configmappropagation/resources/labels.go:
```diff
@@ -0,0 +1,28 @@
+	PropagationLabelValueCopy     = "copy"
+)
+
+func OriginalLabels(selector map[string]string) map[string]string {
```
This function looks quite misleading to me:
1. it mutates the argument in place but returns a map, giving the false impression that a copy is returned
2. it sets something, but doesn't include "Set" in the name

Is it truly necessary?

grac3gao replied:
I'll pass the deepcopy in this method, not the original one (It doesn't need to set something, my mistake.)

antoineco commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -121,62 +126,80 @@ func (r *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error
+		r.Recorder.Event(ns, corev1.EventTypeNormal, configMapPropagationCreated,
+			"Default ConfigMapPropagation: "+cmp.Name+" created")
+		return cmp, nil
+	} else if err != nil {
```
See previous comment about the use of `switch`.

grac3gao replied:
Same here, followed the same way as other methods in this go file did (like reconcileBroker(), reconcileBrokerRBAC())

antoineco commented on cmd/broker/config.go:
```diff
@@ -0,0 +1,39 @@
+// GetLoggingConfig will get config from a specific namespace
+func GetLoggingConfig(ctx context.Context, namespace, loggingConfigMapName string) (*logging.Config, error) {
+	loggingConfigMap, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(namespace).Get(loggingConfigMapName, metav1.GetOptions{})
+	if err != nil {
```
See previous comment about the use of `switch`.

antoineco commented on cmd/broker/config.go:
```diff
@@ -0,0 +1,39 @@
+
+import (
+	"context"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
Maybe separate stdlib imports from the rest with a new line.

antoineco commented on cmd/broker/filter/main.go:
```diff
@@ -76,10 +78,18 @@ func main() {
 	kubeClient := kubeclient.Get(ctx)
 
-	loggingConfig, err := sharedmain.GetLoggingConfig(ctx)
+	loggingConfigMapName := cmpresources.MakeCopyConfigMapName(namespaceresources.DefaultConfigMapPropagationName, logging.ConfigMapName())
```
Shouldn't we have a fallback in case ConfigMapPropagation is disabled (for any reason)?

grac3gao replied:
Could you provide some cases when ConfigMapPropagation will be disabled? And I was wondering what kind of fallback you think we should have (for making the broker reconcile successfully or just for that loggingConfigMapName and metricConfigMapName?)

antoineco replied:
I didn't have any particular scenario for disabling replication in mind, that was a genuine open question because I didn't follow early conversations on that topic. The fallback I was thinking about was to the original hardcoded "kn-eventing/config-xyz" Configmaps, but if config replication becomes a core Kn feature which guarantees a Broker can always expect a copy of the logging and observability configmaps to be present, I'm fine with the implementation :+1:

grac3gao replied:
Thanks! Before this CRD ConfigMapPropagation was introduced, what we used to do is letting broker get the data from original hardcoded ConfigMaps which were in "knative-eventing" namespace (with some RBAC rule). Currently, ConfigMapPropagation only used when a namespace is injecting label "knative-eventing" to start a broker. If the ConfigMapPropagation is disabled, this part of the code in namespace will be modified, and we can set back RBAC rules for the broker and corresponding broker reading ConfigMap logic to use the old way (which would be a fallback as you suggested).

antoineco commented on pkg/tracing/setup.go:
```diff
@@ -81,14 +81,14 @@ func SetupStaticPublishing(logger *zap.SugaredLogger, serviceName string, cfg *t
 // tracing.HTTPSpanMiddleware as a middleware HTTP handler. The configuration will be dynamically
 // updated when the ConfigMap is updated.
-func SetupDynamicPublishing(logger *zap.SugaredLogger, configMapWatcher *configmap.InformedWatcher, serviceName string) error {
+func SetupDynamicPublishing(logger *zap.SugaredLogger, configMapWatcher *configmap.InformedWatcher, serviceName string, tracingConfigName string) error {
```
```suggestion
func SetupDynamicPublishing(logger *zap.SugaredLogger, configMapWatcher *configmap.InformedWatcher, serviceName, tracingConfigName string) error {
```

antoineco commented on pkg/reconciler/testing/configmap.go:
```diff
@@ -0,0 +1,62 @@
+package testing
+
+import (
+	v1 "k8s.io/api/core/v1"
```
Redundant imports alias

antoineco commented on pkg/reconciler/testing/configmap.go:
```diff
@@ -0,0 +1,62 @@
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/eventing/pkg/apis/configs/v1alpha1"
+	"knative.dev/pkg/kmeta"
+
```
Extra new line

antoineco commented on cmd/broker/ingress/main.go:
```diff
@@ -31,9 +31,12 @@ import (
 	"go.opencensus.io/stats/view"
 	"go.uber.org/zap"
 
+	cmdborker "knative.dev/eventing/cmd/broker"
```
Typo "borker" :)

antoineco commented on config/200-addressable-resolvers-clusterrole.yaml:
```diff
@@ -155,3 +155,24 @@ rules:
+    duck.knative.dev/addressable: "true"
+# Do not use this role directly. These rules will be added to the "addressable-resolver" role.
+rules:
+  - apiGroups:
```
The indentation is different from the other roles.

grac3gao commented on pkg/apis/configs/v1alpha1/configmappropagation_types.go:
```diff
@@ -0,0 +1,94 @@
+	// inherits duck/v1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1.Status `json:",inline"`
```
> If I understand the API correctly, you can specify a namespace and a selector which specifies which configmaps are to be propagated?

Yes, it will copy multiple ConfigCaps from a namespace (specified in the `spec.originalnamespace`) with specified selector to current namespace. 

> I think we'd want to know which configmaps were copied and probably some version information (generation?) or something like that.

Currently, there will be a Normal-level Event under this API resource when one ConfigMap copied successfully, showing the name of the original ConfigMap. And there will be a Warn-level Event when one ConfigMap copy failed, showing the original ConfigMap name and the error message. For the copied ConfigMap in current namespace, it will have a ownerreference. This ownerreference points to the configmappropagation which creates this configmap. And the copied configmap’s name follows the rule `configmappropagation name + “-” + original ConfigMap name`” So the ConfigMap information can be checked back implicitly. Those are works have been done related for ConfigMap information. 

I am not sure if those work provide sufficient information...Should I add more component under this API's status struct? (like an array of ObjectReference for each copied ConfigMap)

vaikas replied:
Yeah, so I think the status field should reflect what happened and what didn't happen. In other objects which create composite resources, the state of each "child" resource is reflected in the status field, for example, in the Subscription we don't just say Ready, we reflect dependency statuses in the Status field.
I think as a user, having to dig through the OwnerRefs is not really the most friendly way to figure out what happened.
My preference would be to have an array of each mirrored ConfigMap and maybe even the Generation (maybe overkill?) and that array then turns into "Ready" state if each attempted operation succeeded. I also think that simply saying copy succeeded is not enough, say you start with copying 3 successfully (status is Ready), then one gets deleted from the source, but deletion from the target fails (so you still have 3 though you should only have 2), then I'd argue that this should be a failure, since the mirror operation did not succeed. So, perhaps the array should be:
 - CMPropagation {
      Source: srcNS/scrName
      Op: copy|delete
      Ready: true|false|unknown(?)
      Reason: some string here in error case
}
Or some array like that  so that as a user if the resource does not get Ready, I can figure out what it was trying to do (for example, I don't want to execute the label selector query to see which sources _should_ be even attempted to be copied/removed).
Does that make sense?

grac3gao replied:
Yes, this will be much more user-friendly. Thank you for this detailed and informative suggestion! I'll add this status modification.

lionelvillard replied:
when removing `knative.dev/config-propagation:copy` I get this:

```
NAME               READY   REASON              ORIGINALNAMESPACE   SELECTOR
knative-eventing   False   PropagationFailed   knative-eventing    map[knative.dev/config-category:eventing]
```

Should `ready` be `True` and `Op` be `nocopy` (or something similar). Also `SELECTOR` does not format properly.

grac3gao replied:
> Should `ready` be `True` and `Op` be `nocopy` (or something similar). Also `SELECTOR` does not format properly.

I just replied in https://github.com/knative/eventing/issues/1758#issuecomment-574444026 regarding this 😄  I'll modify it for the case when `knative.dev/config-propagation:copy` removed.

grac3gao commented on config/300-configmappropagation.yaml:
```diff
@@ -0,0 +1,68 @@
+  names:
+    kind: ConfigMapPropagation
+    plural: configmappropagations
+    singular: configmappropagation
```
I checked both eventing and serving, it doesn't conflict with other shortnames. I'll add it.

grac3gao commented on pkg/apis/configs/v1alpha1/configmappropagation_validation.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+// Validate the ConfigMapPropagation.
+func (cmp *ConfigMapPropagation) Validate(ctx context.Context) *apis.FieldError {
```
No, you don't miss anything. It is me who missed to add it...Thank you for pointing out!

grac3gao commented on config/200-addressable-resolvers-clusterrole.yaml:
```diff
@@ -155,3 +155,24 @@ rules:
+  name: configs-addressable-resolver
+  labels:
+    eventing.knative.dev/release: devel
+    duck.knative.dev/addressable: "true"
```
removed

grac3gao commented on pkg/reconciler/configmappropagation/resources/configmap.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+func MakeCopyConfigMapLabel(configMapPropagationNamespace, configMapName string) string {
+	return configMapPropagationNamespace + "-" + configMapName
```
I also feel ` / ` is more proper than `-`. However, when I tried to set label like `knative.dev/config-original: configMapPropagationNamespace/configMapName`, it returned error for invalid value. It seems label keys accept `/`, but label values only accept `-`,`.`,`_`,  according to https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

grantr replied:
You're right! I didn't know about that. Can you add a comment to this method explaining why it uses dash?

yolocs commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,322 @@
+
+func (r *Reconciler) getOriginalLabelSelector(cmp *v1alpha1.ConfigMapPropagation) labels.Selector {
+	return labels.SelectorFromSet(resources.OriginalLabels(r.copies(cmp.Spec.Selector)))
+}
```
nit: This is used only once? Maybe just use a variable where it needs?

yolocs commented on pkg/reconciler/configmappropagation/resources/configmap_test.go:
```diff
@@ -0,0 +1,77 @@
+			})
+			// If name and namespace are correct.
+			if name := configmap.Name; name != "cmp-original-config-map" {
+				t.Errorf("Expected name %q, actually %q", "cmp-original-config-map", name)
```
nit: use -got +want, and other places

grantr commented:
/hold

Hold until after 0.12 cut. I'd like to see this land in early 0.13 so it gets enough exercise before release.

/milestone v0.13.0-M1

chizhg commented:
Can we have an E2E test for it? Or having an issue for tracking if not in this PR?

mattmoor commented on pkg/apis/configs/register.go:
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_defaults.go:
```diff
@@ -0,0 +1,29 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_defaults_test.go:
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle.go:
```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle_test.go:
```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_types.go:
```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_validation.go:
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/configmappropagation_validation_test.go:
```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/doc.go:
```diff
@@ -0,0 +1,20 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/register.go:
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/apis/configs/v1alpha1/register_test.go:
```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/configmappropagation_test.go:
```diff
@@ -0,0 +1,481 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/controller.go:
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/controller_test.go:
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/resources/configmap.go:
```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/resources/configmap_test.go:
```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/resources/labels.go:
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/configmappropagation/resources/labels_test.go:
```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/namespace/resources/configmappropagation.go:
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/testing/configmap.go:
```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/reconciler/testing/configmappropagation.go:
```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2019 The Knative Authors
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

grac3gao commented:
Ready for review. Changes include adding copy configmap status, updating perf test setup, updating of ObservedGeneration, and others.

@vaikas When I tried to set the source ConfigMap's generation under array of each mirrored ConfigMap's status, I found ConfigMaps in knative-eventing don’t have generation. Thus, generation never shows up even I set it. I checked with @grantr. It seems in knative-eventing, ConfigMap doesn't set generation in metadata... So, in this change, I didn't include generation information under status array.

@lionelvillard I have updated the logic for removing copy label, and updated corresponding status. I used operation `stop` together with reason: `copy ConfigMap doesn't have "knative.dev/config-propagation:copy" label, stop propagating this ConfigMap` to indicate this behavior.  For the `SELECTOR` format, the additionalPrinterColumns only supports integer, number, string, boolean, date (https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/#type), so I used string. Is there any way to make it in a better format?

lionelvillard commented on pkg/apis/configs/v1alpha1/configmappropagation_types.go:
```diff
@@ -0,0 +1,113 @@
+	OriginalNamespace string `json:"originalNamespace,omitempty"`
+	// Selector only selects original configMaps with corresponding labels
+	// +optional
+	Selector *map[string]string `json:"selector,omitempty"`
```
what about using LabelSelector no avoid surprises: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#labelselector-v1-meta

antoineco replied:
+1
Also maps _are_ pointers already.

grantr replied:
:+1: to using `*metav1.LabelSelector` as used in [DeploymentSpec](https://github.com/kubernetes/api/blob/9bd6c7846cc9c14a6440b78ad9f66fc36769473e/apps/v1/types.go#L279) and others in the apps group.

grantr replied:
@antoineco In API types, fields like this are pointers to create a distinction between a missing value and an empty value. For LabelSelector specifically this distinction has semantic meaning: ["An empty label selector matches all objects. A null label selector matches no objects."](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#labelselector-v1-meta)

antoineco replied:
@grantr right, but _missing_ values are usually expected to be set by a defaulter (e.g. Replicas, TerminationGracePeriod, ...). Is that the case here?

grantr replied:
I'm only interpreting the docs; I have no inside knowledge of Kubernetes API design. :)

In some cases like the ones you mention, the missing value is a signal to the defaulting webhook that a default should be set, and the controller would not be expected to encounter an object with a missing value.

But in the case of the LabelSelector, a defaulter would not work because replacing a missing value with an empty value would convert the selector's meaning from "no objects" to "all objects" (as documented). The controller is expected to interpret a missing label selector differently from a present-but-empty label selector.

Does this make sense? I could be reading the docs wrong...

antoineco replied:
It does yes, my point was more about the rejection than the defaulting. Maps already have this distinction between nil and empty built into the Go language, that's why I was saying a pointer to a map was error prone.
In the case of a struct, I agree we should absolutely use a pointer.

grantr replied:
Ah I'm sorry, I understand now! Good point about maps already allowing null and empty. ObjectMeta is an example with labels and annotations being (optional) maps:
https://github.com/kubernetes/apimachinery/blob/fb4ce0f5343808fd521a80b5dc63e535c357039f/pkg/apis/meta/v1/types.go#L231

lionelvillard commented on pkg/apis/configs/v1alpha1/configmappropagation_types.go:
```diff
@@ -0,0 +1,113 @@
+	duckv1.Status `json:",inline"`
+
+	//CopyConfigMaps is the status for each copied configmap.
+	CopyConfigMaps map[string]ConfigMapPropagationStatusCopyConfigMap `json:"CopyConfigmaps,omitempty"`
```
should be `copyConfigmaps`. Also it's not very k8s to use a map here. Have you considered using an array? Like `conditions`

grac3gao replied:
I used map because I did't want to append two status with the same source in an array. But after you pointed out `conditions`, I found it does `patchStrategy` and `patchMergeKey` to avoid that. Changed to array. Thank you! 
For the `SELECTOR`, I'll just remove it…

lionelvillard commented:
thanks! I added few comments to the code.

Regarding the format, one idea is to add a new status field, called `selector` of type string and change the CRD to point to it. However selector can be quite complicated so I'm not sure this is useful to print it.

vaikas commented:
@grac3gao re: generation, perhaps we should include ResourceVersion there in that case then? I'm just curious how we detect if things have changed.
https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes

grantr commented:
> I'm just curious how we detect if things have changed.

Are you talking about how the controller detects if things have changed, or the user? Can you explain a bit more how you expect this part of the status to be used?

grac3gao commented:
Ready for review. Changes include change copy configmap status to be array, change selector from map to LabelSelector, adding validation for LabelSelector, adding resourceVersion to configmap status.

lionelvillard commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,343 @@
+			logging.FromContext(ctx).Debug("Propagate ConfigMap " + configMap.Name + " succeed")
+			r.Recorder.Eventf(cmp, corev1.EventTypeNormal, configMapPropagationPropagateSingleConfigMapSucceed,
+				fmt.Sprintf("Propagate ConfigMap %v succeed", configMap.Name))
+			expectedStatus.SetCopyConfigMapStatus(name, source, copyConfigMap, "true", "", resourceVersion)
```
to be consistent it should be `True` instead of `true`

lionelvillard commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,343 @@
+	configMapPropagationPropagateSingleConfigMapSucceed = "ConfigMapPropagationPropagateSingleConfigMapSucceed"
+
+	// Name of operation to propagate single configmap.
+	copyConfigMap   = "copy"
```
should be `Copy` for consistenty sake

lionelvillard commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,343 @@
+
+	// Name of operation to propagate single configmap.
+	copyConfigMap   = "copy"
+	deleteConfigMap = "delete"
```
same here: `Delete`

lionelvillard commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,343 @@
+	copyConfigMap   = "copy"
+	deleteConfigMap = "delete"
+	// stopConfigMap indicates a configmap stop propagating.
+	stopConfigMap = "stop"
```
and here `Stop`

grac3gao commented:
Done.

grantr commented:
/hold cancel

grantr commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle.go:
```diff
@@ -52,3 +52,24 @@ func (cmps *ConfigMapPropagationStatus) MarkNotPropagated() {
 }
+
+func (cmpsc *ConfigMapPropagationStatusCopyConfigMap) SetCopyConfigMapStatus(name, source, operation, ready, reason, resourceVersion string) {
+	if name != "" {
```
What's the point of these guards? It looks like they may be incorrect. They don't clear the field's existing value, but that seems to be the intent of the callers.

grac3gao replied:
You are right, no need for these guards. Originally, I thought twhen updating copy configmap status, it may inherit some attributes from the old configmap status(like source, name), but I didn't use this way later in the reconciler logic. I removed those guards.

grantr commented on pkg/apis/configs/v1alpha1/configmappropagation_lifecycle_test.go:
```diff
@@ -199,3 +200,25 @@ func TestConfigMapPropagationIsReady(t *testing.T) {
+		cmpsc *ConfigMapPropagationStatusCopyConfigMap
+		want  *ConfigMapPropagationStatusCopyConfigMap
+	}{{
+		name:  "all happy",
```
If the empty string guards are kept, a test case should be added that verifies they work.

grantr commented on pkg/apis/configs/v1alpha1/configmappropagation_validation.go:
```diff
@@ -47,31 +41,30 @@ func (cmps *ConfigMapPropagationSpec) Validate(ctx context.Context) *apis.FieldE
+		if cmps.Selector.MatchExpressions != nil {
 			fe := &apis.FieldError{
-				Message: "At least one selector must be specified",
+				Message: fmt.Sprintf("MatchExppressions isn't supported yet"),
```
typo nit: `matchExpressions`

grantr commented on pkg/reconciler/configmappropagation/configmappropagation_test.go:
```diff
@@ -0,0 +1,493 @@
+				Eventf(corev1.EventTypeNormal, configMapPropagationReadinessChanged, "ConfigMapPropagation %q became ready", configMapPropagationName),
+			},
+		}, {
+			Name: "Copy ConfigMap no longer has required labels",
```
Related to the comment above and #2437: if we write code to only delete the CMP owner reference, we should add a test case for abandoning a ConfigMap with multiple owner references.

mattmoor commented on config/200-addressable-resolvers-clusterrole.yaml:
```diff
@@ -1 +0,0 @@
-core/roles/addressable-resolvers-clusterrole.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/roles/addressable-resolvers-clusterrole.yaml

```

mattmoor commented on config/200-controller-clusterrole.yaml:
```diff
@@ -1 +0,0 @@
-core/roles/controller-clusterroles.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/roles/controller-clusterroles.yaml

```

mattmoor commented on config/config-logging.yaml:
```diff
@@ -1 +0,0 @@
-core/configmaps/logging.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/configmaps/logging.yaml

```

mattmoor commented on config/config-observability.yaml:
```diff
@@ -1 +0,0 @@
-core/configmaps/observability.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/configmaps/observability.yaml

```

mattmoor commented on config/config-tracing.yaml:
```diff
@@ -1 +0,0 @@
-core/configmaps/tracing.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/configmaps/tracing.yaml

```

grac3gao commented on pkg/reconciler/configmappropagation/configmappropagation.go:
```diff
@@ -0,0 +1,343 @@
+		cmp.Status.CopyConfigMaps = append(cmp.Status.CopyConfigMaps, expectedStatus)
+	}
+	// List ConfigMaps in current namespace and delete copy ConfigMap if the corresponding original ConfigMap no longer exists or no longer has the required label.
+	copyConfigMapList, err := r.configMapLister.ConfigMaps(cmp.Namespace).List(labels.Everything())
```
Yeah, I listed everything because we used to check copy configmap with copy label removed by users, and returned error. Since we are no longer do that (currently using stop propagating for copy configmap without copy label), we can only return configmaps with copy label.

grac3gao commented:
Thank you for reviewing it and opening subsequent issues! @grantr. I changed api group to be `configs.internal.knative.dev`. Together with modification according to reviews, ready for review.

grantr commented on config/300-configmappropagation.yaml:
```diff
@@ -0,0 +1,62 @@
+# Copyright 2019 The Knative Authors
```
Now that #2428 has merged, this file should be moved to `config/core/resources/configmappropagation.yaml`, with a symlink `config/300-configmappropagation.yaml` pointing there.

grac3gao replied:
done.

mattmoor commented on config/300-configmappropagation.yaml:
```diff
@@ -0,0 +1 @@
+core/resources/configmappropagation.yaml
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
core/resources/configmappropagation.yaml

```

grac3gao replied:
Does symbolic link also need trailing newline?

grantr replied:
No, don't edit this file. @mattmoor your bot should probably ignore symlinks

grantr commented:
/kind feature
/area observability
/area api

grantr commented:
/remove-area test-and-release
